### PR TITLE
power token & mining in ipor router

### PR DESCRIPTION
**Pull request by amydlarz**

### CODE REVIEW

**Review by mario-ipor — CHANGES_REQUESTED** [review]
Names of method are not unique and not describe which module is used, prepare proposition with more descriptive names

File: contracts/interfaces/ILiquidityMiningV2.sol [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+import "./types/LiquidityMiningTypes.sol";
+
+/// @title The interface for interaction with the LiquidityMining.
+/// LiquidityMining is responsible for the distribution of the Power Token rewards to accounts
+/// staking lpTokens and / or delegating Power Tokens to LiquidityMining. LpTokens can be staked directly to the LiquidityMining,
+/// Power Tokens are a staked version of the [Staked] Tokens minted by the PowerToken smart contract.
+interface ILiquidityMiningV2 {
+    /// @notice Contract ID. The keccak-256 hash of "io.ipor.LiquidityMining" decreased by 1
+    /// @return Returns an ID of the contract
+    function getContractId() external pure returns (bytes32);
```
[the remark this answers is not in the document]
Comment (amydlarz) [reply contracts/interfaces/ILiquidityMiningV2.sol:13]: this contract is not used in router

File: contracts/interfaces/ILiquidityMiningV2.sol [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+    /// @notice Returns the balance of staked lpTokens
+    /// @param account the account's address
+    /// @param lpToken the address of lpToken
+    /// @return balance of the lpTokens staked by the sender
+    function balanceOf(address account, address lpToken) external view returns (uint256);
```
[the remark this answers is not in the document]
Comment (amydlarz) [reply contracts/interfaces/ILiquidityMiningV2.sol:19]: like above

File: contracts/interfaces/ILiquidityMiningV2.sol [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    /// @return An array of `AccountRewardResult` structures, containing the LP token address, rewards amount, and allocated Power Token balance for the account.
+    function calculateAccountRewards(address account, address[] calldata lpTokens)
+        external
+        view
+        returns (LiquidityMiningTypes.AccountRewardResult[] memory);
+
+    /// @notice method allowing to update the indicators per asset (lpToken).
+    /// @param account of which we should update the indicators
+    /// @param lpTokens of the staking pools to update the indicators
+    function updateIndicators(address account, address[] calldata lpTokens) external;
```
[the remark this answers is not in the document]
Comment (amydlarz) [reply contracts/interfaces/ILiquidityMiningV2.sol:50]: like above

File: contracts/interfaces/ILiquidityMiningV2.sol [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    /// @notice Removes Power Tokens from the liquidity mining for multiple accounts.
+    /// @param updatePwToken An array of `UpdatePwToken` structures, each containing the account address,
+    /// LP token address, and Power Token amount to be removed.
+    function removePwTokens(LiquidityMiningTypes.UpdatePwToken[] memory updatePwToken) external;
+
+    /// @notice Claims accumulated rewards for multiple LP tokens and transfers them to the specified account.
+    /// @param account The account address to claim rewards for.
+    /// @param lpTokens An array of LP token addresses for which rewards will be claimed.
+    /// @return rewardsAmountToTransfer The total amount of rewards transferred to the account.
+    function claim(address account, address[] calldata lpTokens)
```
[the remark this answers is not in the document]
Comment (amydlarz) [reply contracts/interfaces/ILiquidityMiningV2.sol:76]: like above

File: contracts/interfaces/IFlowsService.sol [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    /// @dev Reverts if the `lpTokens` array is empty.
+    /// @dev Reverts if there are no rewards to claim.
+    function claimPowerToken(address[] calldata lpTokens) external;
+
+    /// @notice Updates the indicators for a given account and LP tokens.
+    /// @param account The account address for which the indicators are to be updated.
+    /// @param lpTokens An array of LP tokens for which the indicators are to be updated.
+    /// @dev This function calls the `updateIndicators` function of the `ILiquidityMiningV2` contract to update the indicators.
+    /// @dev Reverts if the `lpTokens` array is empty.
+    function updateIndicators(address account, address[] calldata lpTokens) external;
```
Comment (mario-ipor) [inline contracts/interfaces/IFlowsService.sol]: Indicators is quite generic name which can correspond to amm module so please add to this method some power token context for example: updateLiquidityMiningIndicators

File: contracts/interfaces/IFlowsService.sol [hunk 6]
```diff
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity 0.8.16;
+
+interface IFlowsService {
```
Comment (mario-ipor) [inline contracts/interfaces/IFlowsService.sol:4]: Maybe better IPowerTokenFlowsService?

File: contracts/router/IporProtocolRouter.sol [hunk 7]
```diff
@@ -92,17 +128,43 @@ contract IporProtocolRouter is UUPSUpgradeable, AccessControl {
             return AMM_SWAPS_LENS;
         } else if (
             sig == IAmmPoolsLens.getPoolConfiguration.selector ||
-            sig == IAmmPoolsLens.getExchangeRate.selector ||
-            sig == IAmmPoolsLens.getBalance.selector ||
+            sig == IAmmPoolsLens.getIpTokenExchangeRate.selector ||
+            sig == IAmmPoolsLens.getIpTokenBalance.selector ||
```
Comment (mario-ipor) [inline contracts/router/IporProtocolRouter.sol:132]: This is not only ipToken balance - better use getAmmBalance()

File: contracts/router/IporProtocolRouter.sol [hunk 8]
```diff
@@ -216,6 +278,29 @@ contract IporProtocolRouter is UUPSUpgradeable, AccessControl {
             _nonReentrant();
             _reentrancyStatus = _ENTERED;
             return AMM_GOVERNANCE_SERVICE_ADDRESS;
+        } else if (
+            sig == IStakeService.stakeLpTokens.selector ||
```
Comment (mario-ipor) [inline contracts/router/IporProtocolRouter.sol:282]: Maybe better stakeLpTokensToLiquidityMining

File: contracts/router/IporProtocolRouter.sol [hunk 9]
```diff
@@ -216,6 +278,29 @@ contract IporProtocolRouter is UUPSUpgradeable, AccessControl {
...
+            sig == IStakeService.unstakeLpTokens.selector ||
```
Comment (mario-ipor) [inline contracts/router/IporProtocolRouter.sol:283]: unstakeLpTokensFromLiqudityMining

File: contracts/router/IporProtocolRouter.sol [hunk 10]
```diff
@@ -216,6 +278,29 @@ contract IporProtocolRouter is UUPSUpgradeable, AccessControl {
...
+            sig == IStakeService.stakeProtocolToken.selector ||
```
Comment (mario-ipor) [inline contracts/router/IporProtocolRouter.sol:284]: stakeIporToken

File: contracts/router/IporProtocolRouter.sol [hunk 11]
```diff
@@ -216,6 +278,29 @@ contract IporProtocolRouter is UUPSUpgradeable, AccessControl {
...
+            sig == IStakeService.unstakeProtocolToken.selector ||
```
Comment (mario-ipor) [inline contracts/router/IporProtocolRouter.sol:285]: unstakeIporToken

File: contracts/interfaces/IAmmPoolsLens.sol [hunk 12]
```diff
@@ -24,18 +24,18 @@ interface IAmmPoolsLens {
...
     /// @param asset The address of the asset.
     /// @return uint256 The ipToken exchange rate for the specific asset, represented in 18 decimals.
-    function getExchangeRate(address asset) external view returns (uint256);
+    function getIpTokenExchangeRate(address asset) external view returns (uint256);
 
     /// @notice Retrieves the AmmTreasury balance for a given asset.
     /// @param asset The address of the asset.
     /// @return IporTypes.AmmBalancesMemory The balance of the AMM Treasury.
-    function getBalance(address asset) external view returns (IporTypes.AmmBalancesMemory memory);
+    function getIpTokenBalance(address asset) external view returns (IporTypes.AmmBalancesMemory memory);
```
Comment (mario-ipor) [inline contracts/interfaces/IAmmPoolsLens.sol:38]: This is not  iptoken balance use getAmmBalance

**Review by mario-ipor — CHANGES_REQUESTED** [review]

File: contracts/interfaces/IFlowsService.sol (hunk 5 above)
Comment (amydlarz) [reply contracts/interfaces/IFlowsService.sol]: I changed it before your comment ;)

File: contracts/interfaces/ILiquidityMiningLens.sol [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+pragma solidity 0.8.16;
+
+import "./types/LiquidityMiningTypes.sol";
+import "./ILiquidityMiningV2.sol";
+
+interface ILiquidityMiningLens {
+
+    /// @notice Contract ID. The keccak-256 hash of "io.ipor.LiquidityMining" decreased by 1
+    /// @return Returns an ID of the contract
+    function getLiquidityMiningContractId() external view returns (bytes32);
```
Comment (amydlarz) [inline contracts/interfaces/ILiquidityMiningLens.sol:11]: delete

File: contracts/interfaces/ILiquidityMiningLens.sol [hunk 14]
```diff
@@ -0,0 +1,61 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity 0.8.16;
+
+import "./types/LiquidityMiningTypes.sol";
+import "./ILiquidityMiningV2.sol";
```
Comment (amydlarz) [inline contracts/interfaces/ILiquidityMiningLens.sol:5]: delete

File: contracts/interfaces/ILiquidityMiningLens.sol [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+    /// @return balances list of {LiquidityMiningTypes.DelegatedPwTokenBalance} structure, with information how much Power Token is delegated per lpToken address.
+    function balanceOfDelegatedPowerToken(address account, address[] memory lpTokens)
+        external
+        view
+        returns (LiquidityMiningTypes.DelegatedPwTokenBalance[] memory balances);
+
+    /// @notice Calculates the accrued rewards for the specified LP tokens in the Liquidity Mining contract.
+    /// @param lpTokens An array of LP tokens for which the accrued rewards are to be calculated.
+    /// @return result An array of `AccruedRewardsResult` structs containing the accrued rewards information for each LP token.
+    function calculateLiquidityMiningAccruedRewards(address[] calldata lpTokens)
```
Comment (amydlarz) [inline contracts/interfaces/ILiquidityMiningLens.sol:31]: getAccruedRewardsInLiquidityMining

File: contracts/interfaces/ILiquidityMiningLens.sol [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+    /// @return An array of `AccountRewardResult` structs containing the rewards information for each LP token.
+    function calculateLiquidityMiningAccountRewards(address account, address[] calldata lpTokens)
+        external
+        view
+        returns (LiquidityMiningTypes.AccountRewardResult[] memory);
+
+    /// @notice Retrieves the global indicators for the specified LP tokens in the Liquidity Mining contract.
+    /// @param lpTokens An array of LP tokens for which the global indicators are to be retrieved.
+    /// @return An array of `GlobalIndicatorsResult` structs containing the global indicators information for each LP token.
+    function getLiquidityMiningGlobalIndicators(address[] memory lpTokens)
```
Comment (amydlarz) [inline contracts/interfaces/ILiquidityMiningLens.sol:48]: getGlobalIndicatorsFromLiquidityMining

File: contracts/interfaces/ILiquidityMiningLens.sol [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+        external
+        view
+        returns (LiquidityMiningTypes.GlobalIndicatorsResult[] memory);
+
+    /// @notice Retrieves the account indicators for the specified account and LP tokens in the Liquidity Mining contract.
+    /// @param account The address of the account for which the account indicators are to be retrieved.
+    /// @param lpTokens An array of LP tokens for which the account indicators are to be retrieved.
+    /// @return An array of `AccountIndicatorsResult` structs containing the account indicators information for each LP token.
+    function getLiquidityMiningAccountIndicators(address account, address[] memory lpTokens)
```
Comment (amydlarz) [inline contracts/interfaces/ILiquidityMiningLens.sol:57]: getAccountIndicatorsFromLiquidityMining

File: contracts/interfaces/IPowerTokenLens.sol [hunk 18]
```diff
@@ -0,0 +1,62 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity 0.8.16;
+import "./types/PowerTokenTypes.sol";
+
+interface IPowerTokenLens {
+
+    /// @notice Gets the name of the Power Token
+    /// @return Returns the name of the Power Token.
+    function powerTokenName() external view returns (string memory);
```
Comment (amydlarz) [inline contracts/interfaces/IPowerTokenLens.sol:9]: delete

File: contracts/interfaces/IPowerTokenLens.sol [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+
+    /// @notice Contract ID. The keccak-256 hash of "io.ipor.PowerToken" decreased by 1
+    /// @return Returns the ID of the contract
+    function getPowerTokenContractId() external view returns (bytes32);
```
Comment (amydlarz) [inline contracts/interfaces/IPowerTokenLens.sol:13]: delete

File: contracts/interfaces/IPowerTokenLens.sol [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+
+    /// @notice Gets the symbol of the Power Token.
+    /// @return Returns the symbol of the Power Token.
+    function powerTokenSymbol() external view returns (string memory);
```
Comment (amydlarz) [inline contracts/interfaces/IPowerTokenLens.sol:17]: delete

File: contracts/interfaces/IPowerTokenLens.sol [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+
+    /// @notice Returns the number of the decimals used by Power Token. By default it's 18 decimals.
+    /// @return Returns the number of decimals: 18.
+    function powerTokenDecimals() external view returns (uint8);
```
Comment (amydlarz) [inline contracts/interfaces/IPowerTokenLens.sol:21]: delete

File: contracts/interfaces/IPowerTokenFlowsService.sol [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+    /// @dev Reverts if the `lpTokens` array is empty.
+    /// @dev Reverts if there are no rewards to claim.
+    function claimPowerToken(address[] calldata lpTokens) external;
+
+    /// @notice Updates the indicators for a given account and LP tokens.
+    /// @param account The account address for which the indicators are to be updated.
+    /// @param lpTokens An array of LP tokens for which the indicators are to be updated.
+    /// @dev This function calls the `updateIndicators` function of the `ILiquidityMiningV2` contract to update the indicators.
+    /// @dev Reverts if the `lpTokens` array is empty.
+    function updateLiquidityMiningIndicators(address account, address[] calldata lpTokens) external;
```
Comment (amydlarz) [inline contracts/interfaces/IPowerTokenFlowsService.sol:18]: updateIndicatorsInLiquidityMining

File: contracts/interfaces/IStakeService.sol [hunk 23]
```diff
@@ -0,0 +1,70 @@
+// SPDX-License-Identifier: BSD-3-Clause
+pragma solidity 0.8.16;
+
+interface IPowerTokenStakeService {
```
Comment (amydlarz) [inline contracts/interfaces/IStakeService.sol:4]: change filename to IPowerTokenStakeService

File: contracts/interfaces/IStakeService.sol [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+    ) external;
+
+    /// @notice Stakes the specified amount of IPOR tokens on behalf of the specified address.
+    /// @param onBehalfOf The address on whose behalf the IPOR tokens will be staked.
+    /// @param iporTokenAmount The amount of IPOR tokens to be staked, represented with 18 decimals.
+    /// @dev The function ensures that the provided `onBehalfOf` address is valid and the `iporTokenAmount` is greater than zero.
+    /// @dev The function calls the `addStakedToken` function of the PowerToken contract to update the staked IPOR tokens.
+    /// @dev Finally, the function transfers the IPOR tokens from the sender to the PowerToken contract for staking.
+    /// @dev Reverts if any of the requirements is not met or if the transfer of IPOR tokens fails.
+    function stakeProtocolToken(address onBehalfOf, uint256 iporTokenAmount) external;
```
Comment (amydlarz) [inline contracts/interfaces/IStakeService.sol:44]: stakeGovernanceTokenToPowerToken

File: contracts/interfaces/IStakeService.sol [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+
+    /// @notice Unstakes the specified amount of IPOR tokens and transfers them to the specified address.
+    /// @param transferTo The address to which the unstaked IPOR tokens will be transferred.
+    /// @param iporTokenAmount The amount of IPOR tokens to be unstaked, represented with 18 decimals.
+    /// @dev The function ensures that the `iporTokenAmount` is greater than zero.
+    /// @dev The function calls the `removeStakedTokenWithFee` function of the PowerToken contract to remove the staked IPOR tokens.
+    /// @dev Finally, the function transfers the corresponding staked token amount to the `transferTo` address.
+    /// @dev Reverts if the `iporTokenAmount` is not greater than zero, or if the transfer of staked tokens fails.
+    function unstakeProtocolToken(address transferTo, uint256 iporTokenAmount) external;
```
Comment (amydlarz) [inline contracts/interfaces/IStakeService.sol:53]: unstakeGovernanceTokenFromPowerToken

File: contracts/interfaces/IStakeService.sol [hunk 26]
```diff
@@ -0,0 +1,70 @@
...
+
+    /// @notice Initiates a cooldown period for the specified amount of Power Tokens.
+    /// @param pwTokenAmount The amount of Power Tokens to be put into cooldown, represented with 18 decimals.
+    /// @dev The function ensures that the `pwTokenAmount` is greater than zero.
+    /// @dev The function calls the `cooldown` function of the PowerToken contract to initiate the cooldown.
+    /// @dev Reverts if the `pwTokenAmount` is not greater than zero.
+    function cooldownPowerToken(uint256 pwTokenAmount) external;
```
Comment (amydlarz) [inline contracts/interfaces/IStakeService.sol:60]: pwTokenCooldown

File: contracts/interfaces/IStakeService.sol [hunk 27]
```diff
@@ -0,0 +1,70 @@
...
+
+    /// @notice Cancels the active cooldown for the sender.
+    /// @dev The function calls the `cancelCooldown` function of the PowerToken contract to cancel the cooldown.
+    function cancelPowerTokenCooldown() external;
```
Comment (amydlarz) [inline contracts/interfaces/IStakeService.sol:64]: pwTokenCancelCooldown

File: contracts/interfaces/IStakeService.sol [hunk 28]
```diff
@@ -0,0 +1,70 @@
...
+
+    /// @notice Redeems Power Tokens and transfers the corresponding Staked Tokens to the specified address.
+    /// @dev The function calls the `redeem` function of the PowerToken contract to redeem Power Tokens.
+    /// @param transferTo The address to which the Staked Tokens will be transferred.
+    function redeemPowerToken(address transferTo) external;
```
Comment (amydlarz) [inline contracts/interfaces/IStakeService.sol:69]: redeemPwToken

File: contracts/interfaces/IAssetManagementLens.sol [hunk 29]
```diff
@@ -30,4 +30,7 @@ interface IAssetManagementLens {
     /// @return uint256 The balance of the account in the Compound protocol.
     function compoundBalanceOf(address asset) external view returns (uint256);
 
+    /// @notice Calculated exchange rate between ivToken and the underlying asset. Asset is specific to AssetManagement's intance (ex. USDC, USDT, DAI, etc.)
```
Comment (amydlarz) [inline contracts/interfaces/IAssetManagementLens.sol:33]: balanceOfAmmTreasury -> balanceOfAmmTreasuryInAssetManagement

File: contracts/interfaces/IAssetManagementLens.sol [hunk 30]
```diff
@@ -30,4 +30,7 @@ interface IAssetManagementLens {
...
+    /// @return Current exchange rate between ivToken and the underlying asset, represented in 18 decimals.
```
Comment (amydlarz) [inline contracts/interfaces/IAssetManagementLens.sol:34]: aaveBalanceOf -> aaveBalanceOfInAssetManagement
compoundBalanceOf -> compoundBalanceOfInAssetManagement

File: contracts/interfaces/IAmmPoolsLens.sol [hunk 31]
```diff
@@ -24,18 +24,18 @@ interface IAmmPoolsLens {
     /// @notice Retrieves the configuration of a specific asset's pool.
     /// @param asset The address of the asset.
     /// @return PoolConfiguration The pool's configuration.
-    function getPoolConfiguration(address asset) external view returns (PoolConfiguration memory);
+    function getAmmPoolsLensConfiguration(address asset) external view returns (PoolConfiguration memory);
```
Comment (amydlarz) [inline contracts/interfaces/IAmmPoolsLens.sol:27]: IAmmPoolsService: rebalance -> rebalanceBetweenAmmTraseryAndAssetManagement

File: contracts/interfaces/IAmmPoolsLens.sol [hunk 32]
```diff
@@ -24,18 +24,18 @@ interface IAmmPoolsLens {
...
 
     /// @notice Calculates the ipToken exchange rate.
     /// @dev The exchange rate is a ratio between the Liquidity Pool Balance and the ipToken's total supply.
     /// @param asset The address of the asset.
     /// @return uint256 The ipToken exchange rate for the specific asset, represented in 18 decimals.
-    function getExchangeRate(address asset) external view returns (uint256);
+    function getIpTokenExchangeRate(address asset) external view returns (uint256);
```
Comment (amydlarz) [inline contracts/interfaces/IAmmPoolsLens.sol:33]: redeemUsdt -> redeemFromAmmPoolUsdt

Comment (amydlarz) [reply contracts/interfaces/IAmmPoolsLens.sol:33]: getPoolConfiguration -> getAmmPoolServiceConfiguration

File: test/utils/builder/LiquidityMiningLensBuilder.sol [hunk 33]
```diff
@@ -0,0 +1,29 @@
...
+
+    BuilderUtils.LiquidityMiningLensData private _builderData;
+
+    address private _owner;
+
+    constructor(address owner) {
+        _owner = owner;
+    }
+
+    function withContractId(bytes32 contractId) public returns (LiquidityMiningLensBuilder) {
```
Comment (mario-ipor) [inline test/utils/builder/LiquidityMiningLensBuilder.sol:17]: what is the reason for this method?

File: test/utils/builder/PowerTokenLensBuilder.sol [hunk 34]
```diff
@@ -0,0 +1,30 @@
...
+
+    BuilderUtils.PowerTokenLensData private _builderData;
+
+    address private _owner;
+
+    constructor(address owner) {
+        _owner = owner;
+    }
+
+    function withName(string memory name) public returns (PowerTokenLensBuilder) {
```
Comment (mario-ipor) [inline test/utils/builder/PowerTokenLensBuilder.sol:18]: What for is this set?

**Review by mario-ipor — APPROVED** [review]

**Pull request merged**